### PR TITLE
WIP: Gitlab support

### PR DESCRIPTION
**Pull request by alexandrst88**
Hi. This is PR, will add Gitlab support, it's not 100% finished, because:
- I'd like to know maintainers thoughts about design that I've provided?
- Also, not variables have a clear meaning for Gitlab provider.
Thanks :)

### CODE REVIEW

File: merge/mergeGitlab.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	client.SetBaseURL(os.Getenv("GITLAB_URL"))
+
+	// OK to merge?
+
+	// (1) Check if the PR is mergeable
+	<-githubLimiter.C
+	pid := fmt.Sprintf("%s/%s", input.Org, input.Repo)
+	truePointer := true
+	mr, _, err := client.MergeRequests.GetMergeRequest(pid, input.PRNumber, &gitlab.GetMergeRequestsOptions{IncludeDivergedCommitsCount: &truePointer})
+	//mr, _, err := client.MergeRequests.GetMergeRequest(pid, 492, nil)
```
Comment (nathanleiby) [inline merge/mergeGitlab.go]: nit: needs cleanup

File: cmd/merge.go [hunk 2]
```diff
@@ -94,7 +94,12 @@ func mergeOneRepo(r initialize.Repo, ctx context.Context) error {
 		RequireReviewApproval: !mergeFlagIgnoreReviewApproval,
 		RequireBuildSuccess:   !mergeFlagIgnoreBuildStatus,
 	}
-	output, err := merge.Merge(ctx, input, githubLimiter, mergeThrottle)
+	var output merge.Output
+	if r.Provider == "gitlab" {
+		output, err = merge.GitlabMerge(input, githubLimiter, mergeThrottle)
+	} else if r.Provider == "github" {
+		output, err = merge.GitHubMerge(ctx, input, githubLimiter, mergeThrottle)
+	}
```
Comment (nathanleiby) [inline cmd/merge.go]: we should error in the case that it's neither of these (same in other if / else if blocks where we check `r.Provider`)

File: cmd/root.go [hunk 3]
```diff
@@ -25,15 +25,16 @@ var rootCmd = &cobra.Command{
...
+	if os.Getenv("GITHUB_API_TOKEN") == "" && os.Getenv("GITLAB_API_TOKEN") == "" {
+		log.Fatalf("GITHUB_API_TOKEN or GITLAB_API_TOKEN env var is not set. In order to use microplane, create a token (https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/) then set the env var.")
 	}
 
 	rootCmd.PersistentFlags().StringP("repo", "r", "", "single repo to operate on")
 	rootCmd.AddCommand(cloneCmd)
 	rootCmd.AddCommand(docsCmd)
 	rootCmd.AddCommand(initCmd)
 
+	initCmd.Flags().StringVarP(&repoProviderFlag, "provider", "p", "", "Git Repository Provider: github or gitlab")
```
Comment (nathanleiby) [inline cmd/root.go]: nit: for readability, group this into a block in the same way done below for the other commands

File: cmd/root.go [hunk 4]
```diff
@@ -25,15 +25,16 @@ var rootCmd = &cobra.Command{
 }
 
 func init() {
-	if os.Getenv("GITHUB_API_TOKEN") == "" {
-		log.Fatalf("GITHUB_API_TOKEN env var is not set. In order to use microplane, create a token (https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/) then set the env var.")
+	if os.Getenv("GITHUB_API_TOKEN") == "" && os.Getenv("GITLAB_API_TOKEN") == "" {
+		log.Fatalf("GITHUB_API_TOKEN or GITLAB_API_TOKEN env var is not set. In order to use microplane, create a token (https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/) then set the env var.")
 	}
```
[the remark this answers is not in the document]
Comment (nathanleiby) [reply cmd/root.go]: also: ideally we could link to gitlab docs if you're expecting gitlab token and missing it

File: cmd/init.go [hunk 5]
```diff
@@ -19,13 +21,22 @@ $ mp init "org:Clever filename:circle.yml"
 would target all Clever repos with a circle.yml file.
 
 See https://help.github.com/articles/searching-code/ for more details about the syntax.`,
```
Comment (nathanleiby) [inline cmd/init.go]: is there any equivalent gitlab URL you'd recommend? e.g. https://docs.gitlab.com/ee/user/search/advanced_global_search.html

File: merge/mergeGitlab.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/xanzy/go-gitlab"
+)
+
+// Merge an open PR in Github
+// - githubLimiter rate limits the # of calls to Github
+// - mergeLimiter rate limits # of merges, to prevent load when submitting builds to CI system
+func GitlabMerge(input Input, githubLimiter *time.Ticker, mergeLimiter *time.Ticker) (Output, error) {
+	// Create Github Client
+	client := gitlab.NewClient(nil, os.Getenv("GITLAB_API_TOKEN"))
+	client.SetBaseURL(os.Getenv("GITLAB_URL"))
```
Comment (nathanleiby) [inline merge/mergeGitlab.go]: should document somewhere that this env var can be set and how to use it

File: merge/mergeGitlab.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+		_, err := client.MergeRequests.RebaseMergeRequest(pid, input.PRNumber)
+		if err != nil {
+			return Output{Success: false}, fmt.Errorf("Failed to rebase from master")
+		}
+	}
+
+	// Merge the PR
+	<-mergeLimiter.C
+	<-githubLimiter.C
+	time.Sleep(20 * time.Second)
```
Comment (nathanleiby) [inline merge/mergeGitlab.go]: ideally, this sleep should not be necessary, and instead the limiters should take care of it. can we remove it?

File: push/pushGitlab.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	return pr, nil
+}
+
+func GetPipelineStatus(client *gitlab.Client, owner string, name string, opts *gitlab.ListProjectPipelinesOptions) (string, error) {
+	pid := fmt.Sprintf("%s/%s", owner, name)
+	pipeline, _, err := client.Pipelines.ListProjectPipelines(pid, opts)
+	if err != nil {
+		return "", errors.New("unexpected: cannot get pipeline status")
+	}
+	return pipeline[0].Status, nil
```
Comment (nathanleiby) [inline push/pushGitlab.go]: we should verify that >0 pipelines are returned, otherwise this will panic

File: cmd/merge.go (hunk 2 above)
Comment (alexandrst88) [reply cmd/merge.go]: I've checking provider here

```		if provider != "github" && provider != "gitlab" {
			log.Fatal("--provider must be github or gitlab")
		}```

and write provider in the state file, in order to not provide provider on future steps as plan, push, merge. But if you want to double-check it, I could provide another condition step

File: cmd/root.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexandrst88) [reply cmd/root.go]: I've trying to figure out the best approach of the provider. It's not ideal, but you can check updated PR. thanks

File: cmd/init.go (hunk 5 above)
Comment (alexandrst88) [reply cmd/init.go]: done

**Comment by nathanleiby** [comment]
Thanks for all the improvements! Merging in.

**Pull request merged**